### PR TITLE
Add information to type checking errors

### PR DESCRIPTION
**Pull request by Cypher1**
Also adds a test that ensures these error messages are generated.

### CODE REVIEW

File: src/runtime/recipe/handle.ts [hunk 1]
```diff
@@ -223,9 +223,14 @@ export class Handle implements Comparable<Handle> {
...
-  static resolveEffectiveType(handleType: Type, connections: HandleConnection[]): Type {
-    const typeSet = connections.filter(connection => connection.type != null).map(connection => ({type: connection.type, direction: connection.direction}));
-    return TypeChecker.processTypeList(handleType, typeSet);
+  static resolveEffectiveType(handleType: Type, connections: HandleConnection[], options: IsValidOptions): Type {
+    const typeSet: TypeListInfo[] = connections.filter(
+      connection => connection.type != null
+    ).map(
+    connection => (
+      {type: connection.type, direction: connection.direction}
+    ));
```
Comment (csilvestrini) [inline src/runtime/recipe/handle.ts]: Line breaks look kind of strange here. I think you can condense these quite a bit.

File: src/runtime/recipe/recipe.ts [hunk 2]
```diff
@@ -33,7 +33,7 @@ export type CloneMap = Map<RecipeComponent, RecipeComponent>;
 export type VariableMap = Map<TypeVariableInfo|Schema, TypeVariableInfo|Schema>;
 
 export type IsResolvedOptions = {showUnresolved?: boolean, details?: string[]}; // TODO(lindner): standardize details
-export type IsValidOptions = {errors?: Map<Recipe | RecipeComponent, string>};
+export type IsValidOptions = {errors?: Map<Recipe | RecipeComponent, string>, typeErrors?: string[]}; // TODO(cypher1): Fixup
```
Comment (csilvestrini) [inline src/runtime/recipe/recipe.ts]: Fix up how?

File: src/runtime/recipe/handle.ts (hunk 1 above)
Comment (shans) [reply src/runtime/recipe/handle.ts]: I agree; either go back to a single line, or
```
const typeSet = connections
    .filter(connection => connection.type != null)
    .map(connection => {type: connection.type, direction: connection.direction}));
```

What you have currently is too hard to read.

File: src/runtime/recipe/type-checker.ts [hunk 3]
```diff
@@ -76,7 +82,7 @@ export class TypeChecker {
         }
         return candidate;
       }
-      return null;
+      return error(`could not guarantee variable ${candidate} meets read requirements ${candidate.canReadSubset} with write guarantees ${candidate.canWriteSuperset}`);
```
Comment (shans) [inline src/runtime/recipe/type-checker.ts]: why does the containing function here (processTypeList) return null sometimes and throw an error sometimes now? If there's a good rationale, make it clear in a comment.

File: src/runtime/recipe/handle.ts (hunk 1 above)
Comment (Cypher1) [reply src/runtime/recipe/handle.ts]: If I can push back a little, if this was written with the equivalent procedural code, I would find it hard to read if it were condensed down to only one or two lines.
I think is one of the problems often cited when pushing against 'point free' style in functional languages.

I think your formatting, Shane, is pretty neat.
Is a single or double indent expected?

File: src/runtime/recipe/recipe.ts (hunk 2 above)
Comment (Cypher1) [reply src/runtime/recipe/recipe.ts]: Thanks, I'd thought I could remove the `typeErrors` field (putting the errors back into the error map when they 'bubbled' up to a place where we knew what RecipeComponent they were associated with, unfortunately the typechecker doesn't always know where in the recipe it is operating so we can't.

I'll remove the comment, I don't think it's achievable.

File: src/runtime/recipe/type-checker.ts (hunk 3 above)
Comment (Cypher1) [reply src/runtime/recipe/type-checker.ts]: Sorry, where is the throw?

This is returning null via a helper function that adds an error message to options' error map.
Maybe I should rename `error` to `returnWithMessage` or something similar?

Comment (Cypher1) [reply src/runtime/recipe/type-checker.ts]: As the error function is currently only used in this one place, I've inlined it.

We can generalize the pattern when necessary.

**Comment by Cypher1** [comment]
Copybara has marked this as 'pending' but already has my `+2`. Going to try a close/open cycle again.

**Pull request merged**